homeassistant/components/device_tracker/unifi.py

```diff
@@ -33,55 +40,87 @@
-    except APIError as ex:
-        _LOGGER.error("Failed to connect to Unifi: %s", ex)
+        result = UnifiScanner(config, ectrl=ectrl)
+    except (ValueError, APIError) as ex:
         hass.components.persistent_notification.create(
             'Failed to connect to Unifi. '
-            'Error: {}<br />'
+            'Error: {} <br />'
             'You will need to restart hass after fixing.'
-            ''.format(ex),
+            .format(ex),
             title=NOTIFICATION_TITLE,
             notification_id=NOTIFICATION_ID)
-        return False
-
-    return UnifiScanner(ctrl)
+    return result
 
 
 class UnifiScanner(DeviceScanner):
     """Provide device_tracker support from Unifi WAP client data."""
 
-    def __init__(self, controller):
-        """Initialize the scanner."""
-        self._controller = controller
+    def __init__(self, config, ectrl=None):
+        """Set up the Unifi device_tracker."""
+        from pyunifi.controller import Controller, APIError
+        _LOGGER.debug("pyunifi loaded")
+        host = config.get(CONF_HOST)
+        username = config.get(CONF_USERNAME)
+        password = config.get(CONF_PASSWORD)
+        site_id = config.get(CONF_SITE_ID)
+        port = config.get(CONF_PORT)
+        verify_ssl = config.get(CONF_VERIFY_SSL)
+        detection_time = config.get(CONF_DETECTION_TIME)
+        api_version = config.get(CONF_API_VERSION)
+
+        if api_version != 'v4' and api_version != 'v5':
+            _LOGGER.critical("Invalid API version")
+            raise ValueError("Invalid API version")
+
+        try:
+            if ectrl is None:
+                ctrl = Controller(host, username, password, port,
+                                  version=api_version,
+                                  site_id=site_id, ssl_verify=verify_ssl)
+            else:
+                _LOGGER.debug("External controller used")
+                ctrl = ectrl(host, username, password, port,
+                             version=api_version, site_id=site_id,
+                             ssl_verify=verify_ssl)
+            _LOGGER.info("Connection to controller successful")
+            self._controller = ctrl
+        except APIError as ex:
+            _LOGGER.critical("Failed to connect to Unifi: %s", ex)
+            raise ex
+
+        self._detection_time = detection_time
         self._update()
 
     def _update(self):
         """Get the clients from the device."""
+        _LOGGER.info("Updating client information")
         from pyunifi.controller import APIError
         try:
             clients = self._controller.get_clients()
+            _LOGGER.debug("Found %d clients", len(clients))
         except APIError as ex:
-            _LOGGER.error("Failed to scan clients: %s", ex)
+            _LOGGER.critical("Failed to scan clients: %s", ex)
             clients = []
 
-        self._clients = {client['mac']: client for client in clients}
+        self._clients = {
```

Comment:
it feels like the entire patch should just be these 5 lines plus any new tests, it's really mostly confusing to change all the rest of the content in there, including where the controller is started. Could this be reset and simplified at a starting point?